src/white.jl

```diff
@@ -0,0 +1,110 @@
+constant. In some applications, `X` is a subset of the regressors in the original model,
+or just the fitted values. This saves degrees of freedom and may give a more powerful test.
+The test statistic is T*R2 where R2 is from the regression of `e^2` on the terms
+mentioned above. Under the null hypothesis it is distributed as `Chisquare(dof)`
+where `dof` is the number of independent terms (not counting the constant), so the null
+is rejected when the test statistic is large enough.
+
+Implements: [`pvalue`](@ref)
+
+# References
+  * H. White, (1980): A heteroskedasticity-consistent covariance matrix estimator and a direct test for heteroskedasticity, Econometrica, 48, 817-838.
+  * T.S. Breusch & A.R. Pagan (1979), A simple test for heteroscedasticity and random coefficient variation, Econometrica, 47, 1287-1294
+  * R. Koenker (1981), A note on studentizing a test for heteroscedasticity, Journal of Econometrics, 17, 107-112
+
+# External links
+  * [White's test on Wikipedia](https://en.wikipedia.org/wiki/White_test)
+"""
+function WhiteTest(X::AbstractVecOrMat{<:Real}, e::AbstractVector{<:Real}, TestType=:White)
+    (n,K) = (size(X,1),size(X,2))    #nobs,nvars
+
+    n == length(e) || throw(DimensionMismatch("inputs must have the same length"))
+
+    if TestType == :Linear
+        z  = X
+        TestDescription = "Breusch-Pagan's test for heteroskedasticity (linear)"
+    elseif TestType == :LinearAndSquares
+        z = [X X.^2]
+        TestDescription = "White's test for heteroskedasticity (linear and squares)"
+    else               #linear, squares and cross-products
+        z = fill(NaN,n,round(Int,K*(K+1)/2))   #loop is easier and quicker than index tricks
+        vv = 1
+        for i = 1:K, j = i:K
+            z[:,vv] = X[:,i].*X[:,j]             #eg. x1*x1, x1*x2, x2*x2
+            vv      = vv + 1
+        end
+        TestDescription = "White's test for heteroskedasticity (linear, squares, cross products)"
+    end
+
+    dof  = rank(z) - 1                         #number of independent regressors in z
+    e2  = e.^2
+    b   = z\e2
+    res = e2 - z*b
+    R2  = 1 - var(res)/var(e2)
+    LM  = n*R2                                 #n*R2 in original papers, could also use n*R2/(1-R2)
+    return WhiteTest(dof, LM, TestDescription)
+end
+
+"""
+    BreuschPaganTest(X, e)
+
+Compute Breusch-Pagan's test for heteroskedasticity.
+
+`X` is the matrix of regressors from the original model and `e` the vector of residuals.
+See `WhiteTest` for further details.
+"""
+BreuschPaganTest(X, e) = WhiteTest(X, e, :Linear)
+
+
+testname(t::WhiteTest)                     = t.TestDescription
+population_param_of_interest(t::WhiteTest) = ("T*R2", 0, t.LM)
+default_tail(test::WhiteTest)              = :right
+
+function show_params(io::IO, t::WhiteTest, ident)
+    println(io, ident, "T*R^2 statistic:        ", t.LM)
+    println(io, ident, "degrees of freedom:     ", t.dof)
```

Comment:
Better print the test type here by storing it in a field, rather than in the description. See how this is done in `KSampleADTest`.

Reply:
Sure, can be done. Either that way of by changing the `testname()` to print the proper heading. The latter is perhaps better?

Reply:
Not sure what "proper" means, but my suggestion was to have `testname` always return the same texte ("White's test for heteroskedasticity"), but print additional information saying which variant is used.